### DISCUSSION TITLE
Computed field

### DISCUSSION THREAD

**Discussion by alexbrowng** [directus/directus#8541] [Feature Requests]
Create a field (alias type) that allows you to create a computed field with the same functionality as the related value display but with the fields of the same item, for example, when you want to display the full_name joining two fields or an address of the type "{{first_name}} {{last_name}} <{{email}}>"

**Comment by rijkvanzanten** [comment]
This is another interesting angle at Views as well @Oreilles. We should consider generated expression column as part of that bigger Views discussion

**Comment by benhaynes** [comment]
Depending on how https://github.com/directus/directus/discussions/3079 goes, this could be combined. My thinking with that other feature was to have functions, like `SUM()`, etc... but we could also have `CONCAT()` for strings!

**Comment by alexbrowng** [comment]
I'm thinking that maybe a regex can be created to convert "{{first_name}} {{last_name}} <{{email}}}>" in CONCAT(collection.first_name, " ", collection.last_name, " <", collection.email, ">")

**Reply by benhaynes** [reply]
No need! We already have a "Template String" option that we can reuse. I think this concat is different enough where it could be its own Display... and the Calculated one (referenced above) can be more for other opertations.

**Reply by rijkvanzanten** [reply]
Lets not confuse display output with actual computed _columns_ in the database 👍🏻

**Reply by alexbrowng** [reply]
The "Template String" is used for a display, but if you want to query this data from the API?

**Reply by rijkvanzanten** [reply]
That's where "calculated" / "computed" fields come in. Those would be actual database _Generated Columns_ which are readonly expression based values 🙂 That's where you'd use a CONCAT() or SUM() etc etc.

**Comment by rothnic** [comment]
Just wanted to follow-up and mention that I just went through 9.0.1 and templated/calculated fields are a key missing piece compared to our custom-built cms.

We have a data entry form that takes in key input values and produces a templated output value. We provide the content editor a preview of the templated output so they know what the current inputs/settings would result in on the frontend. Then, if for whatever reason the templated output doesn't meet their needs, then they can choose to leverage an override field. 

We currently utilize Twig for that functionality and have some rather large templates that would be extremely complicated to implement with CONCAT(). Some are actually producing a chunk of HTML that represents a bulk of the description we are writing. Having these templates in the CMS instead of in the code that renders the content means that the business logic for how to present certain types of information can be managed from the CMS, rather than rolling out new code.

I was just wanting to provide our experience to suggest that a very simple computed field as outlined might not meet the needs of those that heavily use templated fields that need to be exposed via the API.

I'm curious if there are alternative approaches suggested for meeting the same needs.

**Reply by oreilles** [reply]
There is an alternative; you could use [hooks](https://docs.directus.io/extensions/hooks/) to generate your computed field on the fly everytime an item is created/updated:
```ts
function generateColumn(row) {
	row['my_computed_field'] = my_expression();
}

module.exports = function registerHook({ filter }) {
	filter('<my_collection>.items.create', (input) => input.map(generateColumn));
	filter('<my_collection>.items.update', (input) => input.map(generateColumn));
};
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
